File: pb/src/main/java/org/sqlbuilder/pb/objects/Role.java
package org.sqlbuilder.pb.objects;

import org.sqlbuilder.common.*;

import java.io.Serializable;
import java.util.Comparator;
import java.util.Objects;

public class Role implements HasId, Insertable, Comparable<Role>, Serializable
{
	public static final Comparator<Role> COMPARATOR = Comparator //
			.comparing(Role::getRoleSet) //
			.thenComparing(Role::getArgn) //
			.thenComparing(Role::getFunc, Comparator.nullsFirst(Comparator.naturalOrder())) //
			;

	public static final SetCollector<Role> COLLECTOR = new SetCollector<>(COMPARATOR);

	private final RoleSet roleSet;

	private final String argn;

	@Nullable
	private final Theta theta;

	@Nullable
	private final Func func;

	private final String descr;

	// C O N S T R U C T O R

	public static Role make(final RoleSet roleSet, final String n, final String f, final String descriptor, final String theta)
	{
		var r = new Role(roleSet, n, f, descriptor, theta);
		COLLECTOR.add(r);
		return r;
	}

	private Role(final RoleSet roleSet, final String n, final String func, final String descriptor, final String theta)
	{
		this.roleSet = roleSet;
		this.argn = n;
		this.theta = theta == null || theta.isEmpty() ? null : Theta.make(theta);
		this.func = func == null || func.isEmpty() ? null : Func.make(func);
		this.descr = descriptor;
	}

	// A C C E S S

	public RoleSet getRoleSet()
	{
		return this.roleSet;
	}

	public String getArgn()
	{
		return this.argn;
	}

	public Func getFunc()
	{
		return func;
	}

	public Theta getTheta()
	{
		return theta;
	}

	public String getDescr()
	{
		return this.descr;
	}

	@RequiresIdFrom(type = Role.class)
	@Override
	public Integer getIntId()
	{
		return COLLECTOR.get(this);
	}

	@RequiresIdFrom(type = Role.class)
	public static Integer getIntId(final Role role)
	{
		return role == null ? null : COLLECTOR.get(role);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Role that = (Role) o;
		return roleSet.equals(that.roleSet) && argn.equals(that.argn) && Objects.equals(func, that.func);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(roleSet, argn, func);
	}

	// O R D E R I N G

	@Override
	public int compareTo(@NotNull final Role that)
	{
		return COMPARATOR.compare(this, that);
	}

	// I N S E R T

	@RequiresIdFrom(type = RoleSet.class)
	@RequiresIdFrom(type = Func.class)
	@RequiresIdFrom(type = Theta.class)
	@Override
	public String dataRow()
	{
		// (roleid),narg,theta,func,roledescr,rolesetid
		return String.format("'%s',%s,%s,%s,%d", //
				argn, //
				Utils.nullable(theta, HasId::getSqlId), //
				Utils.nullable(func, HasId::getSqlId), //
				Utils.nullableQuotedEscapedString(descr), //
				roleSet.getIntId() //
		);
	}

	@Override
	public String comment()
	{
		return String.format("%s,%s,%s", roleSet.getName(), theta != null ? theta.getTheta() : "∅", func != null ? func.getFunc() : "∅");
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		if (this.descr == null)
		{
			return String.format("%s[%s-%s]", roleSet, argn, func);
		}
		return String.format("%s[%s-%s '%s']", roleSet, argn, func, descr);
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/objects/Role.kt
package org.sqlbuilder.pb.objects

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.*
import org.sqlbuilder.pb.foreign.AliasFnFeLinks
import org.sqlbuilder.pb.foreign.AliasRoleLinks
import org.sqlbuilder.pb.foreign.AliasVnRoleLinks
import java.io.Serializable
import java.util.*

class Role private constructor(
    val roleSet: RoleSet,
    val argType: String,
    func: String,
    descriptor: String,
    vnLinks: Collection<String>?,
    fnLinks: Collection<String>?,
) : HasId, Insertable, Comparable<Role>, Serializable {

    private val func: Func = Func.make(func)

    private val descr: String = descriptor

    // role names for VerbNet and FrameNet
    val aliasVnRoleLinks: AliasRoleLinks? = if (vnLinks == null || vnLinks.isEmpty()) null else AliasVnRoleLinks.make(vnLinks)
    val aliasFnFeLinks: AliasRoleLinks? = if (fnLinks == null || fnLinks.isEmpty()) null else AliasFnFeLinks.make(fnLinks)

    // N I D

    @RequiresIdFrom(type = Role::class)
    override fun getIntId(): Int {
        return COLLECTOR.apply(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Role
        return roleSet == that.roleSet && argType == that.argType && func == that.func
    }

    override fun hashCode(): Int {
        return Objects.hash(roleSet, argType, func)
    }

    // O R D E R I N G

    override fun compareTo(@NotNull that: Role): Int {
        return COMPARATOR.compare(this, that)
    }

    // I N S E R T

    @RequiresIdFrom(type = RoleSet::class)
    @RequiresIdFrom(type = Func::class)
    @RequiresIdFrom(type = AliasRoleLinks::class)
    override fun dataRow(): String {
        return "'$argType',${func.sqlId},${Utils.nullable(aliasVnRoleLinks) { it!!.sqlId }},${Utils.nullable(aliasFnFeLinks) { it!!.sqlId }},${Utils.quotedEscapedString(descr)},${roleSet.intId}"
    }

    override fun comment(): String {
        return "$func, ${roleSet.name}, ${aliasVnRoleLinks?.names ?: "∅"}, ${aliasFnFeLinks?.names ?: "∅"}"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "$roleSet[$argType-$func '$descr']"
    }

    companion object {

        val COMPARATOR: Comparator<Role> = Comparator
            .comparing<Role, RoleSet> { it.roleSet }
            .thenComparing<String> { it.argType }
            .thenComparing<Func>({ it.func }, Comparator.nullsFirst<Func>(Comparator.naturalOrder<Func>()))

        @JvmField
        val COLLECTOR = SetCollector<Role>(COMPARATOR)

        fun make(roleSet: RoleSet, n: String, f: String, descriptor: String, vnLinks: Collection<String>?, fnLinks: Collection<String>?): Role {
            val r = Role(roleSet, n, f, descriptor, vnLinks, fnLinks)
            COLLECTOR.add(r)
            return r
        }

        @Suppress("unused")
        @RequiresIdFrom(type = Role::class)
        fun getIntId(role: Role): Int {
            return COLLECTOR.apply(role)
        }
    }
}
